### PR TITLE
span, relation, download examples

### PR DESCRIPTION
**Pull request by leonardlin**
# Description

Download project and examples based on 1.7+ API
Ability to filter for confirmed examples
Added span, span-types, relations, relations-type for NER labeling

Fixes # (issue)
#72

### CODE REVIEW

**Comment by leonardlin** [comment]
@ljades sorry, I know it's not good practice to have such big PRs, but it's been a while since I made those changes.

**Comment by ljades** [comment]
> @ljades sorry, I know it's not good practice to have such big PRs, but it's been a while since I made those changes.

All good, thank you so much for doing this! Will review it soon

**Comment by leonardlin** [comment]
will fix the linting errors and need to update some test cases, will update pr soon.

File: doccano_api_client/beta/README.md [hunk 1]
```diff
@@ -27,26 +27,13 @@ my_project_name_controller.examples.create(my_example)
...
-```shell
-pipenv install
-pipenv install --dev
-```
-
-- Adjust the credentials in ""./doccano_api_client/beta/tests/test_full_integration_tests.py"
-  - DOCCANO_ENDPOINT = "http://localhost:8000"  # TODO: Make this a pytest parameter
-  - DOCCANO_USER = "admin"
-  - DOCCANO_PASS = "password"
-- Swith to the root of the beta client eg. ./doccano_api_client/beta/
```
Comment (ljades) [inline doccano_api_client/beta/README.md:48]: The instructions here for parametrizing the integration test credentials has been removed, but it is still needed right now, no?

File: doccano_api_client/beta/client.py [hunk 2]
```diff
@@ -16,6 +16,7 @@ def __init__(self, base_url: str) -> None:
         headers = {
             "content-type": "application/json",
             "accept": "application/json",
+            "referer": base_url,
```
Comment (ljades) [inline doccano_api_client/beta/client.py:19]: For my own curiosity, what is the purpose of this addition? Was the API changed so this is now needed?

File: doccano_api_client/beta/controllers/example.py [hunk 3]
```diff
@@ -88,27 +94,16 @@ def get(self, example_id: int) -> ExampleController:
...
-        #       it's ordered by ID. For randomized order, it's ordered randomly but with a fixed
-        #       seed (the requesting user's id).
-        while len(example_dicts) != 0:
-            # Given a single page, iterate through the dicts on that page
-            for example_dict in example_dicts:
-                # TODO (louis): make a Example.from_json() method to handle this. Also, do this
-                #               for all other models
+        response = self.client_session.get(f"{self.examples_url}?confirmed={confirmed}")
+
+        while True:
```
Comment (ljades) [inline doccano_api_client/beta/controllers/example.py:101]: You may disagree, and let me know if you do, but I'm not a fan of using `while True` here if avoidable. Based on the use of `break` down below, you can probably rewrite this without needing `while True` fairly easily.

File: doccano_api_client/beta/controllers/project.py [hunk 4]
```diff
@@ -75,25 +118,34 @@ def get(self, project_id: int) -> ProjectController:
...
-            sanitized_project_dict = {
-                proj_key: project_dict[proj_key] for proj_key in project_obj_fields
-            }
-
-            yield ProjectController(
-                project=Project(**sanitized_project_dict),
-                id=project_dict["id"],
-                projects_url=self.projects_url,
-                client_session=self.client_session,
+        while True:
```
Comment (ljades) [inline doccano_api_client/beta/controllers/project.py:122]: Same with the `while True` over here

File: doccano_api_client/beta/models/examples.py [hunk 5]
```diff
@@ -1,20 +1,18 @@
...
 
-from . import Annotation
-
 
 @dataclass
 class Example(DataClassJsonMixin):
     """Contains the data and operations relevant to a example on a Doccano project"""
 
-    text: str
+    text: str = ""
```
Comment (ljades) [inline doccano_api_client/beta/models/examples.py:11]: Can an example text be completely empty? Do we want to support that?

File: doccano_api_client/beta/tests/controllers/test_example.py [hunk 6]
```diff
@@ -48,44 +47,44 @@ def test_controller_urls(self):
...
-            annotation_mocks.annotation_create_json["label"],
-        )
-
-    @responses.activate
-    def test_create_annotations_with_bad_response(self) -> None:
-        responses.add(bad.bad_post_response)
-        with self.assertRaises(DoccanoAPIError):
-            example = Example(text="blablabla", annotations=[CategoryAnnotation(label=1, prob=0)])
-            self.examples_controller.create(example)
+    #    @responses.activate
```
Comment (ljades) [inline doccano_api_client/beta/tests/controllers/test_example.py:50]: What is the reason for all these commented out test cases?

**Review by ljades — CHANGES_REQUESTED** [review]
Awesome job so far. This is a long PR but it'll be useful for getting up to date with the next version. A few extra things:

1. While you don't have to write as exhaustive tests as I wrote for starters, since there are so many new controllers and components you added, it would be super helpful for tech debt removal and stability if you could write at least some test cases for the controllers for relation, relation_type, span, and span_type. Also, provide reason or replacement for the removed example controller tests.
2. Since there are so many new types, can you add some more documentation to the README? Doesn't need to be too extensive, just provide more examples of the client being used as intended keeping all these controllers in mind.
3. Please update the setup.py as well at: https://github.com/doccano/doccano-client/blob/master/setup.py#L25 . Update where the comment says what versions it is compatible with.
4. Update in the README here as well: https://github.com/doccano/doccano-client/blame/master/doccano_api_client/beta/README.md#L5

Thanks again for getting on this!

**Comment by ljades** [comment]
Oh I just remembered you had mentioned you were gonna update test cases soon, I forgot about that. Yep, take your time!

**Pull request merged**